lib/ansible/parsing/vault/__init__.py

```diff
@@ -526,32 +526,59 @@ def write_data(self, data, filename, shred=True):
 
         if filename == '-':
             sys.stdout.write(b_file_data)
-        else:
-            if os.path.isfile(filename):
-                if shred:
-                    self._shred_file(filename)
-                else:
-                    os.remove(filename)
-            with open(filename, "wb") as fh:
-                fh.write(b_file_data)
+            return
+
+        target_filename, _ = self._unlink_file(filename, shred=shred)
+
+        with open(target_filename, "wb") as fh:
+            fh.write(b_file_data)
+
+    def _unlink_file(self, filename, shred=True):
+        '''unlink a file taking symlinks into account.
+
+        Returns the 'real' filename and its stat info.
+        For symlinks, the returned info will be for the
+        target of the symlink.
+
+        This will unlink the target of a symlink, not the
+        symlink itself.
+
+        If the filename can't be found, the stat info
+        returned will be None.
+        '''
+        target_filename = filename
+        stat_info = None
+        try:
+            target_filename = os.readlink(filename)
+        except OSError:
+            # if readlink fails, assume its not a symlink
+            pass
+
+        if os.path.isfile(target_filename):
+            stat_info = os.stat(target_filename)
+            # old file 'dest' was encrypted, no need to _shred_file
+            if shred:
+                self._shred_file(target_filename)
+            else:
+                os.remove(target_filename)
+
+        return target_filename, stat_info
 
     def shuffle_files(self, src, dest):
         prev = None
+
         # overwrite dest with src
-        if os.path.isfile(dest):
-            prev = os.stat(dest)
-            # old file 'dest' was encrypted, no need to _shred_file
-            os.remove(dest)
-        shutil.move(src, dest)
+        target_dest, prev = self._unlink_file(dest)
```

Comment:
this creates race condition now, there is a time between remove and move that the file is unavailable